src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -1057,39 +1102,55 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger? errorLogger, Cancella
-            if (HasUnsuppressableErrors(diagnostics))
-            {
-                return;
-            }
 
             DiagnosticBag? analyzerExceptionDiagnostics = null;
-            if (!analyzers.IsEmpty || !generators.IsEmpty)
+
+            // If there are parsing errors, we want to return immediately.
+            // But first, we need to check two things: 
+            // 1. Whether there are any suppressible warnings,
+            // 2. Whether there are any diagnostic suppressors that could potentially suppress them.
+            // If both conditions are true, run diagnostic suppressors before exiting from this method.
+            if (HasUnsuppressableErrors(diagnostics))
             {
-                var analyzerConfigProvider = CompilerAnalyzerConfigOptionsProvider.Empty;
-                if (Arguments.AnalyzerConfigPaths.Length > 0)
+                if (HasSuppressableWarningsOrErrors(diagnostics) && analyzers.Any(a => a is DiagnosticSuppressor))
                 {
-                    Debug.Assert(analyzerConfigSet is object);
-                    analyzerConfigProvider = analyzerConfigProvider.WithGlobalOptions(new DictionaryAnalyzerConfigOptions(analyzerConfigSet.GetOptionsForSourcePath(string.Empty).AnalyzerOptions));
+                    var analyzerConfigProvider = GetCompilerAnalyzerConfigOptionsProvider(analyzerConfigSet, additionalTextFiles, diagnostics, compilation, sourceFileAnalyzerConfigOptions);
 
-                    // TODO(https://github.com/dotnet/roslyn/issues/31916): The compiler currently doesn't support
-                    // configuring diagnostic reporting on additional text files individually.
-                    ImmutableArray<AnalyzerConfigOptionsResult> additionalFileAnalyzerOptions =
-                        additionalTextFiles.SelectAsArray(f => analyzerConfigSet.GetOptionsForSourcePath(f.Path));
+                    AnalyzerOptions analyzerOptions = CreateAnalyzerOptions(additionalTextFiles, analyzerConfigProvider);
 
-                    foreach (var result in additionalFileAnalyzerOptions)
-                    {
-                        diagnostics.AddRange(result.Diagnostics);
-                    }
+                    analyzerCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                    analyzerExceptionDiagnostics = new DiagnosticBag();
+
+                    // PERF: Avoid executing analyzers that report only Hidden and/or Info diagnostics, which don't appear in the build output.
+                    //  1. Always filter out 'Hidden' analyzer diagnostics in build.
+                    //  2. Filter out 'Info' analyzer diagnostics if they are not required to be logged in errorlog.
+                    var severityFilter = SeverityFilter.Hidden;
+
+                    if (Arguments.ErrorLogPath == null)
+                        severityFilter |= SeverityFilter.Info;
```

Comment:
This block of code seems to be copied from another place in this file - can we create a helper method to compute severity filter and share this code?

Reply:
Actually it seems all the code starting from `analyzerCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);` until and including `analyzerDriver = AnalyzerDriver.CreateAndAttachToCompilation(` call that initializes the driver seems to be duplicated. Is it possible to create a local function, say `initializeAnalyzerDriver` that shares all of this code?